pydantic/_internal/_generics.py

```diff
@@ -0,0 +1,283 @@
+            return type_
+        return resolved_list
+
+    if isinstance(type_, PydanticForwardRef):
+        # queue the replacement as a deferred action
+        return type_.replace_types(type_map)
+
+    # If all else fails, we try to resolve the type directly and otherwise just
+    # return the input with no modifications.
+    return type_map.get(type_, type_)
+
+
+def check_parameters_count(cls: type[BaseModel], parameters: tuple[Any, ...]) -> None:
+    actual = len(parameters)
+    expected = len(cls.__pydantic_generic_parameters__ or ())
+    if actual != expected:
+        description = 'many' if actual > expected else 'few'
+        raise TypeError(f'Too {description} parameters for {cls}; actual {actual}, expected {expected}')
+
+
+_visit_counts_context: ContextVar[dict[str, int] | None] = ContextVar('_visit_counts_context', default=None)
+
+
+@contextmanager
+def generic_recursion_self_type(origin: type[BaseModel], args: tuple[Any, ...]) -> Iterator[PydanticForwardRef | None]:
+    """
+    This contextmanager should be placed around recursive calls used to build a generic type,
+    and accept as arguments the generic origin type and the type arguments being passed to it.
+
+    If the same origin and arguments are observed twice, it implies that a self-reference placeholder
+    can be used while building the core schema, and will produce a schema_ref that will be valid in the
+    final parent schema.
+
+    I believe the main reason that the same origin/args must be observed twice is that a BaseModel's
+    inner_schema will be a TypedDictSchema that doesn't include the first occurrence of the PydanticForwardRef
+    reference, so the referenced schema may not end up in the final core_schema unless you expand two
+    layers deep.
+    """
+    visit_counts_by_ref = _visit_counts_context.get()
+    if visit_counts_by_ref is None:
+        visit_counts_by_ref = defaultdict(int)
+        token = _visit_counts_context.set(visit_counts_by_ref)
+    else:
+        token = None
+
+    type_ref = get_type_ref(origin, args_override=args)
+    if visit_counts_by_ref[type_ref] >= 2:
+        self_type = PydanticForwardRef(
+            core_schema.definition_reference_schema(type_ref), origin, ({'kind': 'class_getitem', 'item': args},)
+        )
+        yield self_type
+    else:
+        visit_counts_by_ref[type_ref] += 1
+        yield None
+    if token:
+        _visit_counts_context.reset(token)
+
+
+def recursively_defined_type_refs() -> set[str]:
+    return set((_visit_counts_context.get() or {}).keys())
+
+
+_GENERIC_TYPES_CACHE = GenericTypesCache()
+
+
+def _cache_key(cls: type[BaseModel], _params: Any) -> GenericTypesCacheKey:
+    # TODO: What is happening with args here?
+    #   Note: get_args always returns () if _params is a tuple, and
+    #   if _params is a type, _params and (_params,) should have the same meaning as arguments to __class_getitem__.
+    #   So something seems suspicious..
```

Comment:
@samuelcolvin I'd be interested in any insight you might have into what I'm bringing up in this comment